server/render.js

```diff
@@ -82,7 +82,7 @@ async function doRender (req, res, pathname, query, {
   if (isResSent(res)) return
 
   const renderPage = (enhancer = Page => Page) => {
-    const app = createElement(App, {
+    const app = createElement(enhancer(App), {
```

Comment:
Hmmm this seems very dangerous, the HOC is wrapped twice instead of once. First in App, then in Component 🤔

Reply:
Great catch though, I guess we can remove the enhancer from `Component`, but I'm not 100% sure.

Reply:
@timneutkens I agree that it is dangerous to use the HOC twice, but there are some problems related to each approach:

#### Wrap both `App` and `Component`

Possibly incorrect behaviour due to injecting the `enhancer` HOC twice, very much so dependent on what is injected and would need to be documented at the very least.

#### Wrap only the `App` component

This has implications beyond just this issue, because the default `App` doesn't pass down its props to its children. So HOCs wrapping pages that rely on `enhancer` would stop working. To fix it, the default `App` component would need to pass down more props which seems counterintuitive when `url` was just deprecated and the component is being simplified for customisation. This would also be a breaking change unless default `App` handles it gracefully.

#### Wrap only `App` when it is customised

This is a middle ground that I **think** should work pretty well. Wrap `App` if it is customised and `Component` if it is not. The same caveat as above applies, but that can be solved by documenting this behaviour so people realise they might need to pass down more props in `App`.

Reply:
```

Reply:
also `enhancer` is not documented in the readme and the only use case it was meant to support is styled-components / other css-in-js libs that require the react tree

Reply:
What if a second enhancer was introduced specifically for the app component? Something like `renderApp` could provide a solution for the original issue without changing the existing `renderPage` behavior, which I _think_ provides an easier path to deprecate it if there's no longer a use case.

Reply:
> also enhancer is not documented in the readme and the only use case it was meant to support is styled-components / other css-in-js libs that require the react tree

Didn't even realise it was undocumented and the use cases so narrow, clears things up. But what I am getting at is, if someone is using i.e. `material-ui-next` to wrap their pages directly without an `_app.js` component, won't their code break after this change? Those solutions rely on a `pageContext` prop to be passed down to the styling HOC from the enhancer, and that won't be there anymore.

> Also, it seems highly unlikely to me that someone would use the enhancer to pass down props at the moment, it's counter intuitive and only rendered on the server side

Well yes, but css-in-js solutions do it 😆Although this change should provide an upgrade path for them anyway, unless someone actually has a reason to use both `App` **and** different styling HOCs per-page, which seems rather unlikely.

> What if a second enhancer was introduced specifically for the app component? Something like renderApp could provide a solution for the original issue without changing the existing renderPage behavior, which I think provides an easier path to deprecate it if there's no longer a use case.

Feels like a pretty radical path to take. @timneutkens thoughts?

Reply:
Im implementing @timneutkens suggestion since he knows the library best but am slightly concerned about whether this will break existing styling solutions.